# Task 6.1E-commerce 2020 cleaning and consistency checks

## 1 import and load data
## 2 cleaning
### 2a missing 
### 2b duplicates 
## 3 consistency


In [1]:
# import libraries

import pandas as pd
import numpy as np
import os

In [2]:
#load data
#unicode decode error, using encoding=latin1 
df = pd.read_csv(r"C:\Users\andre\OneDrive\Documents\coursework 2022\CF\Immersion\Achievement 6\2 data\original us ecommerce 2020\US  E-commerce records 2020.csv", encoding = 'latin1')

## 2 cleaning


In [3]:
#check head, tail, shape
df

Order Date  Row ID        Order ID       Ship Mode Customer ID  \
0      01-01-20     849  CA-2017-107503  Standard Class    GA-14725   
1      01-01-20    4010  CA-2017-144463  Standard Class    SC-20725   
2      01-01-20    6683  CA-2017-154466     First Class    DP-13390   
3      01-01-20    8070  CA-2017-151750  Standard Class    JM-15250   
4      01-01-20    8071  CA-2017-151750  Standard Class    JM-15250   
...         ...     ...             ...             ...         ...   
3307   30-12-20     908  CA-2017-143259  Standard Class    PO-18865   
3308   30-12-20     909  CA-2017-143259  Standard Class    PO-18865   
3309   30-12-20    1297  CA-2017-115427  Standard Class    EB-13975   
3310   30-12-20    1298  CA-2017-115427  Standard Class    EB-13975   
3311   30-12-20    5092  CA-2017-156720  Standard Class    JM-15580   

          Segment        Country           City       State  Postal Code  \
0        Consumer  United States         Lorain        Ohio        44052   
1        Consumer  United States    Los Angeles  California        90036   
2     Home Office  United States       Franklin   Wisconsin        53132   
3        Consumer  United States     Huntsville       Texas        77340   
4        Consumer  United States     Huntsville       Texas        77340   
...           ...            ...            ...         ...          ...   
3307     Consumer  United States  New York City    New York        10009   
3308     Consumer  United States  New York City    New York        10009   
3309    Corporate  United States      Fairfield  California        94533   
3310    Corporate  United States      Fairfield  California        94533   
3311     Consumer  United States       Loveland    Colorado        80538   

       Region       Product ID         Category Sub-Category  \
0        East  FUR-FU-10003878        Furniture  Furnishings   
1        West  FUR-FU-10001215        Furniture  Furnishings   
2     Central  OFF-BI-10002012  Office Supplies      Binders   
3     Central  OFF-ST-10002743  Office Supplies      Storage   
4     Central  FUR-FU-10002116        Furniture  Furnishings   
...       ...              ...              ...          ...   
3307     East  TEC-PH-10004774       Technology       Phones   
3308     East  OFF-BI-10003684  Office Supplies      Binders   
3309     West  OFF-BI-10002103  Office Supplies      Binders   
3310     West  OFF-BI-10004632  Office Supplies      Binders   
3311     West  OFF-FA-10003472  Office Supplies    Fasteners   

                                           Product Name    Sales  Quantity  \
0                    Linden 10" Round Wall Clock, Black   48.896         4   
1     Howard Miller 11-1/2" Diameter Brentwood Wall ...  474.430        11   
2               Wilson Jones Easy Flow II Sheet Lifters    3.600         2   
3                         SAFCO Boltless Steel Shelving  454.560         5   
4     Tenex Carpeted, Granite-Look or Clear Contempo...  141.420         5   
...                                                 ...      ...       ...   
3307                          Gear Head AU3700S Headset   90.930         7   
3308               Wilson Jones Legal Size Ring Binders   52.776         3   
3309    Cardinal Slant-D Ring Binder, Heavy Gauge Vinyl   13.904         2   
3310                                 GBC Binding covers   20.720         2   
3311                                Bagged Rubber Bands    3.024         3   

      Discount    Profit  
0          0.2    8.5568  
1          0.0  199.2606  
2          0.0    1.7280  
3          0.2 -107.9580  
4          0.6 -187.3815  
...        ...       ...  
3307       0.0    2.7279  
3308       0.2   19.7910  
3309       0.2    4.5188  
3310       0.2    6.4750  
3311       0.2   -0.6048  

[3312 rows x 19 columns]

In [4]:
# data types 
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3312 entries, 0 to 3311
Data columns (total 19 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Order Date    3312 non-null   object 
 1   Row ID        3312 non-null   int64  
 2   Order ID      3312 non-null   object 
 3   Ship Mode     3312 non-null   object 
 4   Customer ID   3312 non-null   object 
 5   Segment       3312 non-null   object 
 6   Country       3312 non-null   object 
 7   City          3312 non-null   object 
 8   State         3312 non-null   object 
 9   Postal Code   3312 non-null   int64  
 10  Region        3312 non-null   object 
 11  Product ID    3312 non-null   object 
 12  Category      3312 non-null   object 
 13  Sub-Category  3312 non-null   object 
 14  Product Name  3312 non-null   object 
 15  Sales         3312 non-null   float64
 16  Quantity      3312 non-null   int64  
 17  Discount      3312 non-null   float64
 18  Profit        3312 non-null 

In [5]:
#change Order Date data type for analysis
df['Order Date'] = df['Order Date'].astype('datetime64')

In [6]:
# descriptive stats 
df.describe()

Row ID   Postal Code         Sales     Quantity     Discount  \
count  3312.000000   3312.000000   3312.000000  3312.000000  3312.000000   
mean   5087.107488  56186.515097    221.381418     3.766908     0.156467   
std    2817.482266  31980.375516    585.257531     2.221776     0.207429   
min      13.000000   1841.000000      0.444000     1.000000     0.000000   
25%    2655.750000  27978.750000     17.018000     2.000000     0.000000   
50%    5183.500000  60472.500000     53.810000     3.000000     0.200000   
75%    7498.250000  90032.000000    205.105700     5.000000     0.200000   
max    9994.000000  99301.000000  13999.960000    14.000000     0.800000   

            Profit  
count  3312.000000  
mean     28.212340  
std     241.864342  
min   -3839.990400  
25%       1.763200  
50%       8.296800  
75%      28.315125  
max    6719.980800

#### sales min needs to be explored
#### profit min needs to be explored and max could be outlier

### 2a missing values

In [7]:
#verify count of null values
df.isnull().sum()

Order Date      0
Row ID          0
Order ID        0
Ship Mode       0
Customer ID     0
Segment         0
Country         0
City            0
State           0
Postal Code     0
Region          0
Product ID      0
Category        0
Sub-Category    0
Product Name    0
Sales           0
Quantity        0
Discount        0
Profit          0
dtype: int64

In [8]:
#check for dups
df.duplicated().sum()

0

### 2 cleaning summary:
3312 rows 
19 columns
'Order Date' data type changed to datetime 
no missing values
no duplicates


## 3 consistency 

In [9]:
#checking value spread 
df['Order Date'].value_counts()

2020-02-09    36
2020-02-12    34
2020-01-12    34
2020-09-12    33
2020-12-11    30
              ..
2020-04-02     1
2020-06-12     1
2020-03-29     1
2020-03-14     1
2020-12-07     1
Name: Order Date, Length: 322, dtype: int64

In [10]:
(df['Order Date'].value_counts() > 1).sum()

306

In [11]:
(df['Order Date'].value_counts() > 10).sum()

135

#### frequencies in order date column show some potential for trend analysis

In [12]:
#row id count unique values
df['Row ID'].duplicated().sum()

0

#### row id is only unique column in table 

In [13]:
#checking for duplicates to see if used as key
df['Order ID'].value_counts()

CA-2017-100111    14
CA-2017-157987    12
CA-2017-140949     9
CA-2017-117457     9
CA-2017-140872     8
                  ..
CA-2017-145443     1
CA-2017-115175     1
CA-2017-158344     1
CA-2017-163321     1
CA-2017-156720     1
Name: Order ID, Length: 1687, dtype: int64

#### shows duplicates, some orders have more than 1 item 

In [14]:
#checking for uniqueness and consistency in qualitative columns 
df['Ship Mode'].value_counts()

Standard Class    1897
Second Class       657
First Class        572
Same Day           186
Name: Ship Mode, dtype: int64

In [15]:
#checking for uniqueness and consistency in qualitative columns 
df['Ship Mode'].value_counts()

Standard Class    1897
Second Class       657
First Class        572
Same Day           186
Name: Ship Mode, dtype: int64

In [16]:
#checking for uniqueness and consistency in qualitative columns 
df['Ship Mode'].value_counts()

Standard Class    1897
Second Class       657
First Class        572
Same Day           186
Name: Ship Mode, dtype: int64

In [17]:
#checking for uniqueness and consistency in qualitative columns 
df['Customer ID'].value_counts()

SV-20365    20
JL-15835    20
MH-18115    19
Dp-13240    19
LC-16870    17
            ..
LW-16825     1
NW-18400     1
TN-21040     1
BF-10975     1
JM-15580     1
Name: Customer ID, Length: 693, dtype: int64

In [18]:
#repeat the value counting function for rest of columns 
column_names = ['Segment', 'Country', 'City', 'State', 'Postal Code', 'Region', 'Product ID', 'Category', 'Sub-Category', 'Product Name', 'Sales', 'Quantity', 'Discount', 'Profit']
for column in column_names:
        print(df[column].value_counts())

Consumer       1668
Corporate       980
Home Office     664
Name: Segment, dtype: int64
United States    3312
Name: Country, dtype: int64
New York City    306
Los Angeles      210
San Francisco    190
Seattle          182
Philadelphia     182
                ... 
Boise              1
Lebanon            1
Hesperia           1
Kissimmee          1
East Orange        1
Name: City, Length: 350, dtype: int64
California              663
New York                352
Texas                   317
Washington              215
Pennsylvania            197
Illinois                172
Ohio                    161
Florida                 126
North Carolina           85
Tennessee                81
Arizona                  70
Michigan                 67
Georgia                  62
Colorado                 61
Indiana                  61
Kentucky                 51
New Jersey               43
Virginia                 43
Massachusetts            39
Minnesota                38
Maryland                 37
Conne

### 3 consistency summary
- all values are within a range thats reasonable for the dataset so no outliers are removed.
- text is consistent in all but product name column where its unstructured and requires extensive reformatting to analyze, will keep for now but may delete.

